packages/ai-code-completion/src/browser/ai-code-completion-provider.ts

```diff
@@ -0,0 +1,84 @@
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import * as monaco from '@theia/monaco-editor-core';
+
+import { CodeCompletionAgent } from '../common/code-completion-agent';
+import { injectable, inject } from '@theia/core/shared/inversify';
+import { PreferenceService } from '@theia/core/lib/browser';
+import { CancellationTokenSource } from '@theia/core';
+import { PREF_AI_CODE_COMPLETION_PRECOMPUTE } from './ai-code-completion-preference';
+
+interface WithArgs<T = unknown[]> {
+    args: T;
+}
+const hasArgs = <T>(object: {}): object is WithArgs<T> => 'args' in object && Array.isArray(object['args']);
+
+@injectable()
+export class AICodeCompletionProvider implements monaco.languages.CompletionItemProvider {
+
+    @inject(CodeCompletionAgent)
+    protected readonly agent: CodeCompletionAgent;
+
+    @inject(PreferenceService)
+    protected readonly preferenceService: PreferenceService;
+
+    async provideCompletionItems(model: monaco.editor.ITextModel, position: monaco.Position,
+        context: monaco.languages.CompletionContext, token: monaco.CancellationToken): Promise<monaco.languages.CompletionList | undefined> {
+        if (!this.preferenceService.get(PREF_AI_CODE_COMPLETION_PRECOMPUTE, false)) {
+            const result = {
+                suggestions: [{
+                    label: 'AI Code Completion',
+                    detail: 'computes after trigger',
+                    kind: monaco.languages.CompletionItemKind.Text,
+                    insertText: '',
+                    range: {
+                        startLineNumber: position.lineNumber,
+                        startColumn: position.column,
+                        endLineNumber: position.lineNumber,
+                        endColumn: position.column
+                    },
+                    args: []
+                }]
+            };
+            (result.suggestions[0] as WithArgs).args = [...arguments];
+            return result;
+        }
+        const cancellationTokenSource = new CancellationTokenSource();
+        token.onCancellationRequested(() => { cancellationTokenSource.cancel(); });
+        return this.agent.provideCompletionItems(model, position, context, cancellationTokenSource.token);
+    }
+
+    async resolveCompletionItem(item: monaco.languages.CompletionItem, token: monaco.CancellationToken): Promise<monaco.languages.CompletionItem> {
+        if (!hasArgs<Parameters<CodeCompletionAgent['provideCompletionItems']>>(item)) {
```

Comment:
The `args` are set when the precompute preference is off, see the check in `provideCompletionItems`.

The idea of "precompute off" is to still offer the AI code completion as an item, but only really trigger it when the item is selected. Once it's selected the framework will call `resolveCompletionItem` and we use the previously stored args to perform the code completion.

With "precompute on" we directly perform the code completion however it might take a while because of the LLM roundtrip. Then we skip the `resolve` part, as we don't need it.